### PR TITLE
add complete validation for position

### PR DESCRIPTION
**Pull request by taiki1frsh**
Issues: #405 #399 part of #392
Add comprehensive validation for opening position

TODO:
issue: #447

### CODE REVIEW

**Comment by taiki1frsh** [comment]
@mkXultra 
You added this validation, right?
https://github.com/UnUniFi/chain/blob/6fad634cc016bc62d38a01a161829f5b13b193fc/x/derivatives/keeper/perpetual_futures.go#L77

Since I made isValid()  method for Position type checks margin condition, can I remove this checking?

**Comment by mkXultra** [comment]
@taiki1frsh 
sure

**Review by KimuraYu45z — APPROVED** [review]

File: x/derivatives/types/positions.go [hunk 1]
```diff
@@ -15,16 +15,38 @@ type PositionInstance interface {
...
+
+	if !pfPosition.IsValidPositionSize() {
+		return fmt.Errorf("position size is not valid")
 	}
+
 	return nil
 }
 
+func (m Position) IsValidMarginAsset() bool {
+	return !(m.Market.BaseDenom != m.RemainingMargin.Denom && m.Market.QuoteDenom != m.RemainingMargin.Denom)
```
Comment (KimuraYu45z) [inline x/derivatives/types/positions.go:41]: Oh plz wait.
@mkXultra Is this ok?

Comment (mkXultra) [reply x/derivatives/types/positions.go:41]: @taiki1frsh 
is this need condition?
```
m.Market.QuoteDenom != m.RemainingMargin.Denom
```

Comment (taiki1frsh) [reply x/derivatives/types/positions.go]: @mkXultra 
Yes, as far as I know.
I fixed the expression and added a comment above.
Plz check it again.

File: x/derivatives/types/positions.go [hunk 2]
```diff
@@ -15,16 +15,39 @@ type PositionInstance interface {
...
+	if !pfPosition.IsValidPositionSize() {
+		return fmt.Errorf("position size is not valid")
 	}
+
 	return nil
 }
 
+// Margin asset must be one of the market assets.
+func (m Position) IsValidMarginAsset() bool {
+	return (m.Market.BaseDenom == m.RemainingMargin.Denom || m.Market.QuoteDenom == m.RemainingMargin.Denom)
```
Comment (mkXultra) [inline x/derivatives/types/positions.go:42]: LGTM

**Pull request merged**